### PR TITLE
places our Collections class inside of a namespace, called cpp_collections for now.

### PR DESCRIPTION
**Pull request by ianzapolsky**
see #19. refactors range, as well as zip and zipWith, to no longer require the verbose declaration (i.e. Collection<T>::range(x)).

### CODE REVIEW

**Comment by ianzapolsky** [comment]
@natebrennand looking for your thoughts on the new `range` functions specifically.

**Comment by natebrennand** [comment]
looking it over now.
FYI, you can view the PR w/o whitespace changes by appending [`?w=1`](https://github.com/natebrennand/cpp_collections/pull/32/files?w=1)

File: src/collections.h [hunk 1]
```diff
@@ -2,580 +2,581 @@
...
-
-        int end = start + indices[i];
-        thread_data td = { &Data, func, start, end};
-        thread_data_pool.push_back(td);
-        start = end;
+    // Return Collection of numeric types over the range [0, size)
+    template<typename T>
+    Collection<T>&
+    range(T size) {
+        static_assert(std::is_arithmetic<T>::value,
```
Comment (natebrennand) [inline src/collections.h]: slick

File: src/collections.h [hunk 2]
```diff
@@ -2,580 +2,581 @@
...
-        thread_pool[i] = pid;
-
-        int end = start + indices[i];
-        thread_data td = { &Data, func, start, end};
-        thread_data_pool.push_back(td);
-        start = end;
+    // Return Collection of numeric types over the range [0, size)
+    template<typename T>
+    Collection<T>&
+    range(T size) {
```
Comment (natebrennand) [inline src/collections.h]: I think range should still take in an `int`.

**Comment by ianzapolsky** [comment]
also check out the tests: benchmarks seem to be failing, though I can't tell what's going on. they pass fine on my machine

**Comment by natebrennand** [comment]
weird, restarting the tests

File: src/collections.h (hunk 2 above)
Comment (ianzapolsky) [inline src/collections.h]: So if range takes an int, how can we allow users to create ranges of other arithmetic types (double, float, etc)? The only other way I can think of (without using some kind of macro) is giving the function itself a template, so for a range of floats you would write range<float>(1,11). 

But that seems ugly to me

Comment (natebrennand) [inline src/collections.h]: That seems acceptably verbose to me.

-----

For me, this looks much more reasonable

```cpp
Collection<int32> l = range(20);
```

 than

```cpp
auto l = range(int32(20));
```

The latter is a kinda obtuse API imo

Comment (ianzapolsky) [inline src/collections.h]: But your first example cannot compile. you would have to write:
```
Collection<int32> l = range<int32>(20);
```
There is no way for range to infer the type of the collection here

Comment (ianzapolsky) [inline src/collections.h]: But I do agree that your first example is the ideal

Comment (natebrennand) [inline src/collections.h]: So let's leave it as is for now, but I'm going to open an issue to explore this strategy later: http://stackoverflow.com/questions/25375118/template-parameter-based-on-expected-return-type

Comment (ianzapolsky) [inline src/collections.h]: ohhh. i like that

**Comment by ianzapolsky** [comment]
good to merge? wanted to get this out there before your stream PR

**Comment by natebrennand** [comment]
we need to try to resolve the benchmarks failing I think. Looking into causes now

**Comment by natebrennand** [comment]
So I can't reproduce any of them locally and running to through valgrind doesn't indicate anything either. I'm kinda thinking we need to maybe stop running benchmarks in CI. Idk how to go about debugging this

**Comment by ianzapolsky** [comment]
i was having the same thought. imo the benchmarks don't 100% belong in the testing suite anyways

**Comment by ianzapolsky** [comment]
want to remove benchmarks from .travis.yml and merge?

**Comment by natebrennand** [comment]
Ya, the performance of traivs CI  isn't consistent anyways.
just post benchmarks in PR's that concern performance

**Pull request merged**